File: DSA_python/DSA02031 GHÉP CHỮ CÁI.py
def submit():
    c = input()
    n = ord(c)-ord('A')+1
    store = [0]*n
    placed = [False]*n

    def is_consonant(a) -> bool:
        return a == 'A' or a == 'E'
    
    def check(A):
        for i in range(1, len(A)-1):
            if A[i] == 'E' or A[i] == 'A':
                if not (is_consonant(A[i-1]) or is_consonant(A[i+1])):
                    return False
        return True
    
    def recursion(store, placed, j):
        if j == n:
            if check(store):
                print("".join(map(str, store)))
            return
        
        for i in range(n):
            if placed[i] == False:
                store[j] = chr(ord('A')+i)
                placed[i] = True

                recursion(store, placed, j+1)
                placed[i] = False


    recursion(store, placed, 0)



submit()

File: DSA/DSA02031 GHÉP CHỮ CÁI.cpp
#include <bits/stdc++.h>
using namespace std;

int n;

bool is_consonant(char a)
{
    return a == 'A' || a == 'E';
}

bool check(vector<char> &A)
{
    for (int i = 1; i < A.size()-1; i++)
    {
        if (A[i] == 'E' || A[i] == 'A')
        {
            if (!is_consonant(A[i-1]) && !is_consonant(A[i+1]))
            {
                return false;
            }
        }
    }
    return true;
}

void recursion(vector<char> store, vector<bool> &placed, int j)
{
    for (int i = 0; i < n; i++)
    {
        if (!placed[i])
        {
            store[j] = 'A'+i;
            placed[i] = true;

            if (j == n-1)
            {
                if (check(store))
                {
                    for (auto &&k : store)
                    {
                        cout << char(k);
                    }
                    cout << endl;
                }
            }
            else
            {
                recursion(store, placed, j+1);
            }
            placed[i] = false;
        }
    }
}

void submit()
{
    char c;
    cin >> c;
    n = c-'A'+1;
    vector<char> store(n);
    vector<bool> placed(n, false);
    
    recursion(store, placed, 0);
    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    submit();

    return 0;
}